### PR TITLE
tor-hs: add new package

### PR DESCRIPTION
**Pull request by ja-pa**
Maintiner: me
Compile tested: Turris Omnia (TOS5), OpenWrt master
Run tested: Turris Omnia (TOS5), OpenWrt master

Description:
This PR adds package tor-hs which simplifies creating of Tor hidden services on the router.
The main use case is the situation when you need to access remotely to your router from the internet or to other devices on your local network (like IP camera etc.).

Thanks to tor-hs you can generate hidden service for different devices or applications ( web interface, ssh, etc). 

Some advantages:

- onion address is self-authenticating (because it's generated based on  hidden service public key)
- The connection between hidden service and the user is End-to-end encrypted
- There is no need for public IP or nat redirection (you just have to be able to connect to Tor network)
- onion address is completely private and only known to you and whoever you choose to disclose it to

https://2019.www.torproject.org/docs/onion-services

### CODE REVIEW

File: net/tor-hs/files/nextcloud-update.sh [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+			nextcloud_cli config:system:set trusted_domains 1000 --value=$onion
+		fi
+	fi
+}
+
+print_help() {
+	echo "Help"
+}
+
+if [ ! -f "$NEXTCLOUD_CLI_SCRIPT" ]; then
```
Comment (neheb) [inline net/tor-hs/files/nextcloud-update.sh]: I assume this can be replaced with [ -f ] || {  }

File: net/tor-hs/files/nextcloud-update.sh [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+NEXTCLOUD_CLI_SCRIPT="/srv/www/nextcloud/occ"
+
+nextcloud_cli() {
+	sudo -u nobody php-cli "$NEXTCLOUD_CLI_SCRIPT" "$@"
+}
+
+
+nextcloud_add_domain() {
+	onion="$1"
+	if [ ! -z "$onion" ]; then
```
Comment (neheb) [inline net/tor-hs/files/nextcloud-update.sh]: I assume this can be replaced with [ -n ] && { }

**Comment by neheb** [comment]
I find it weird that there are nextcloud references with netcloud not being available.

File: net/tor-hs/files/nextcloud-update.sh (hunk 1 above)
Comment (ja-pa) [reply net/tor-hs/files/nextcloud-update.sh]: Fixed.

File: net/tor-hs/files/nextcloud-update.sh (hunk 2 above)
Comment (ja-pa) [reply net/tor-hs/files/nextcloud-update.sh]: Fixed

**Comment by ja-pa** [comment]
> I find it weird that there are nextcloud references with nextcloud not being available.

Yes, I understand that it's weird but I just want to make some example for Hook Script option in tor-hs. Nextcloud was one of the service which I had installed and needed such an approach.

File: net/tor-hs/Makefile [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+PKG_NAME:=tor-hs
+PKG_VERSION:=0.0.1
+PKG_RELEASE:=1
+
+PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)
+
+PKG_MAINTAINER:=Jan Pavlinec <jan.pavlinec@nic.cz>
+PKG_LICENSE:=GPL-3.0-or-later
+
+PKG_INSTALL:=0
```
Comment (neheb) [inline net/tor-hs/Makefile:19]: Does this need to be explicit?

File: net/tor-hs/Makefile [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+# See /LICENSE for more information.
+#
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=tor-hs
+PKG_VERSION:=0.0.1
+PKG_RELEASE:=1
+
+PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)
```
Comment (neheb) [inline net/tor-hs/Makefile:14]: Is this needed?

File: net/tor-hs/Makefile (hunk 3 above)
Comment (ja-pa) [reply net/tor-hs/Makefile]: Removed. Thx.

File: net/tor-hs/Makefile (hunk 4 above)
Comment (ja-pa) [reply net/tor-hs/Makefile]: Removed. Thx.

**Comment by neheb** [comment]
These scripts should be ran through shellcheck I think.

File: net/tor-hs/files/tor-hs.init [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+		# call hook script
+		$hook_script "--update-onion" "$hostname_uri"
+	fi
+}
+
+parse_common_conf() {
+	local hs_dir generated_config
+	config_get generated_config "common" GenConf
+	config_get hs_dir "common" HSDir
+	[ ! -z "$hs_dir" ] && HS_DIR_PATH="$hs_dir"
```
Comment (neheb) [inline net/tor-hs/files/tor-hs.init:96]: ! -z --> -n

File: net/tor-hs/files/tor-hs.init [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	local config="$1"
+
+	config_get enable_hs "$config" Enabled 0
+	config_get hook_script "$config" HookScript
+	config_get name "$config" Name
+
+	hostname="$HS_DIR_PATH/$name/hostname"
+
+	# check if we should run hook_script
+	if [ "$enable_hs" == "true" ] && [ -x "$hook_script" ] && [ -f "$hostname" ] ; then
```
Comment (neheb) [inline net/tor-hs/files/tor-hs.init:85]: == is a bashism.

File: net/tor-hs/files/tor-hs.init [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+			echo "Info. Not updating tor configuration"
+		else
+			#uci add_list
+			echo "Info. Updating tor configuration"
+			uci add_list tor.conf.tail_include="$TORRC_FILE"
+			uci commit tor
+		fi
+	fi
+
+	if [ $restart_tor == "1" ]; then
```
Comment (neheb) [inline net/tor-hs/files/tor-hs.init:36]: == is a bashism.

File: net/tor-hs/files/tor-hs.init (hunk 6 above)
Comment (ja-pa) [reply net/tor-hs/files/tor-hs.init]: Fixed

File: net/tor-hs/files/tor-hs.init (hunk 7 above)
Comment (ja-pa) [reply net/tor-hs/files/tor-hs.init]: Fixed

File: net/tor-hs/files/tor-hs.init (hunk 5 above)
Comment (ja-pa) [reply net/tor-hs/files/tor-hs.init]: Fixed

**Comment by ja-pa** [comment]
@neheb Thx for review. I fixed script with shellcheck.

**Pull request merged**